### PR TITLE
checker: avoid unnecessary remove disconnected peer with multi orphan peers

### PR DESCRIPTION
**Pull request by lhy1024**
### What problem does this PR solve?

Issue Number: Close #7249

When there are many orphan peers, we don't think disconnected peer are healthy
And when we decide to remove peer, we will pick disconnected peer firstly.

### What is changed and how does it work?

```commit-message
```

### Check List

Tests

- Unit test
- Integration test
image

### Release note

```release-note
None.
```

### CODE REVIEW

**Review by nolouch — APPROVED** [review]

File: pkg/schedule/checker/rule_checker.go [hunk 1]
```diff
@@ -534,8 +534,7 @@ func (c *RuleChecker) fixOrphanPeers(region *core.RegionInfo, fit *placement.Reg
 					return operator.CreatePromoteLearnerOperatorAndRemovePeer("replace-down-peer-with-orphan-peer", c.cluster, region, orphanPeer, pinDownPeer)
 				case orphanPeerRole == metapb.PeerRole_Voter && destRole == metapb.PeerRole_Learner:
 					return operator.CreateDemoteLearnerOperatorAndRemovePeer("replace-down-peer-with-orphan-peer", c.cluster, region, orphanPeer, pinDownPeer)
-				case orphanPeerRole == metapb.PeerRole_Voter && destRole == metapb.PeerRole_Voter &&
-					isDisconnectedPeer(pinDownPeer) && !dstStore.IsDisconnected():
+				case orphanPeerRole == destRole && isDisconnectedPeer(pinDownPeer) && !dstStore.IsDisconnected():
```
Comment (lhy1024) [inline pkg/schedule/checker/rule_checker.go:537]: allow replace learner

File: pkg/schedule/checker/rule_checker.go [hunk 2]
```diff
@@ -549,18 +548,24 @@ func (c *RuleChecker) fixOrphanPeers(region *core.RegionInfo, fit *placement.Reg
...
-				ruleCheckerRemoveOrphanPeerCounter.Inc()
-				return operator.CreateRemovePeerOperator("remove-disconnected-orphan-peer", c.cluster, 0, region, orphanPeer.StoreId)
+				// we need to be careful to remove peer, so we will not think disconnected peer is healthy.
+				disconnectedPeer = orphanPeer
+				continue
 			}
 			if hasHealthPeer {
 				// there already exists a healthy orphan peer, so we can remove other orphan Peers.
 				ruleCheckerRemoveOrphanPeerCounter.Inc()
+				// if there exists a disconnected orphan peer, we will pick it to remove firstly.
```
Comment (lhy1024) [inline pkg/schedule/checker/rule_checker.go:565]: avoid to remove normal peer

**Review by nolouch — APPROVED** [review]

**Comment by lhy1024** [comment]
@rleungx PTAL

File: pkg/schedule/checker/rule_checker.go (hunk 2 above)
Comment (rleungx) [reply pkg/schedule/checker/rule_checker.go:569]: Consider that we have 3 orphan peer, two healthy and one disconnected. Is it possiable that we remove a healthy peer first?

Comment (rleungx) [reply pkg/schedule/checker/rule_checker.go:569]: Should we always remove disconnected peer first?

Comment (lhy1024) [reply pkg/schedule/checker/rule_checker.go:570]: > Should we always remove disconnected peer first?
I think so.

**Review by rleungx — APPROVED** [review]

**Comment by lhy1024** [comment]
image

succesfully run test


image
no failure in tidb

image

all orphan peers are removed with 11min down store time config

**Pull request merged**